### ISSUE TITLE
Add HindiCausalLM: A specialized Hindi language model with grouped query attention (~125M parameters)

### ISSUE THREAD

**Issue by NandhaKishorM** [huggingface/transformers#37835]
### Model description

HindiCausalLM model, a causal language model specifically designed for Hindi text generation. The model was developed by ConvaiInnovations and the original implementation is available on the [Hugging Face Hub](https://huggingface.co/convaiinnovations/hindi-causal-lm).

HindiCausalLM addresses the need for specialized language models for the Hindi language, which is one of the most widely spoken languages in the world with over 600 million speakers. The model provides improved performance on Hindi text generation tasks compared to multilingual models of similar size.

## Model Architecture and Technical Specifications

HindiCausalLM is based on a decoder-only transformer architecture with the following specifications:

- **Base Architecture**: Transformer decoder-only, similar to LLaMA but with architecture modifications
- **Size**: 12 layers, 768 hidden dimensions, ~125M parameters
- **Attention Mechanism**: Implements Grouped Query Attention (GQA) with 16 attention heads and 4 key-value heads for efficient inference
- **Feed-Forward Network**: Uses intermediate size of 3072 with SiLU activation functions
- **Position Encoding**: Rotary Position Embeddings (RoPE) with a base frequency of 10000.0
- **Normalization**: Layer normalization applied before attention and feed-forward blocks (pre-normalization)
- **Vocabulary Size**: 16,000 tokens using a SentencePiece tokenizer optimized for Hindi text
- **Context Window**: Supports sequences up to 512 tokens in length

### Basic Usage for Text Generation

```python
from transformers import HindiCausalLMForCausalLM, HindiCausalLMTokenizer

# Load model and tokenizer
model = HindiCausalLMForCausalLM.from_pretrained("convaiinnovations/hindi-causal-lm")
tokenizer = HindiCausalLMTokenizer.from_pretrained("convaiinnovations/hindi-causal-lm")

# Generate text
input_text = "भारत एक विशाल देश है"  # "India is a vast country"
input_ids = tokenizer(input_text, return_tensors="pt").input_ids

# Generate
outputs = model.generate(
    input_ids, 
    max_length=50, 
    num_return_sequences=1, 
    temperature=0.7, 
    top_p=0.9,
    do_sample=True
)

# Decode the generated text
generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)
print(generated_text)
```

### Classification Example

```python
from transformers import HindiCausalLMForSequenceClassification, HindiCausalLMTokenizer
import torch

# Load model and tokenizer
model = HindiCausalLMForSequenceClassification.from_pretrained(
    "convaiinnovations/hindi-causal-lm", num_labels=3
)
tokenizer = HindiCausalLMTokenizer.from_pretrained("convaiinnovations/hindi-causal-lm")

# Example texts
texts = ["यह एक सकारात्मक समीक्षा है", "यह एक नकारात्मक समीक्षा है", "यह एक तटस्थ समीक्षा है"]
inputs = tokenizer(texts, padding=True, return_tensors="pt")

# Forward pass
with torch.no_grad():
    outputs = model(**inputs)
    predictions = torch.argmax(outputs.logits, dim=1)

print(f"Predictions: {predictions}")
```

### TensorFlow Example

```python
from transformers import TFHindiCausalLMForCausalLM, HindiCausalLMTokenizer
import tensorflow as tf

# Load model and tokenizer
model = TFHindiCausalLMForCausalLM.from_pretrained("convaiinnovations/hindi-causal-lm")
tokenizer = HindiCausalLMTokenizer.from_pretrained("convaiinnovations/hindi-causal-lm")

# Process input
input_text = "हिंदी भाषा बहुत समृद्ध है"  # "Hindi language is very rich"
inputs = tokenizer(input_text, return_tensors="tf")

# Generate
generated_ids = model.generate(inputs.input_ids, max_length=30)
generated_text = tokenizer.decode(generated_ids[0], skip_special_tokens=True)
print(generated_text)
```

### Open source status

- [x] The model implementation is available
- [x] The model weights are available

### Provide useful links for the implementation

https://huggingface.co/convaiinnovations/hindi-causal-lm

**Comment by NandhaKishorM** [comment]
Inference colab notebook at [Colab_Link](https://colab.research.google.com/drive/1kWjqTXDu0ah_yi2LdsHTwDBjFB0KDh2Y?usp=sharing)

**Comment by qgallouedec** [comment]
Hi @NandhaKishorM, apologies for the very late reply, we're catching up on the new-model backlog.

New architectures follow our "Model on the Hub first" path, see and the good news is your model already works: it ships `convaicausallm_model.py` with the `custom_code` tag, so anyone can load it today with `trust_remote_code=True`. No core change needed.

Architecturally `ConvaiCausalLM` is a standard Llama-style decoder (grouped-query attention + RoPE + SiLU MLP, just LayerNorm instead of RMSNorm), so it doesn't add a capability the library lacks, and the current signal is thin (8 downloads / 0 likes on the Hub). For those reasons I'm going to close this as won't-implement for now, but this isn't a no: if the model gains real community traction we'd happily review a native port contributed as a `modular_hindicausallm.py`. Thanks for building for Hindi!
